### PR TITLE
Add CI build

### PR DESCRIPTION
**Pull request by AlexanderRichert-NOAA**
## Description

This PR adds a GCC11-and-MPICH-based CI workflow for building fv3atm, on top of which unit tests can be added. Dependencies are installed through Spack, where the installation directory is cached between runs for a significant speed up.

### Issue(s) addressed

Addresses #729

## Testing

Tested standalone build on personal machine and GH Actions. Built under ufs-weather-model on personal machine.

## Dependencies

none

### CODE REVIEW

File: ci/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,63 @@
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/CMakeModules/Modules)
```
Comment (edwardhartnett) [inline ci/CMakeLists.txt:8]: Can you put comments up at the top of this file?

File: ci/spack.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+spack:
```
Comment (edwardhartnett) [inline ci/spack.yaml:12]: Can you put some comments?

Also did we decide to always use yml not yaml? Or does spack want yaml?

File: io/module_wrt_grid_comp.F90 [hunk 3]
```diff
@@ -29,7 +29,7 @@ module module_wrt_grid_comp
       use mpi
       use esmf
       use fms
-      use mpp_mod, only : mpp_init   ! needed for fms 2023.02
+      use mpp_mod, only : mpp_init, uppercase  ! needed for fms 2023.02
```
Comment (edwardhartnett) [inline io/module_wrt_grid_comp.F90]: Is this change intentional? It doesn't seem related to the rest...

Comment (AlexanderRichert-NOAA) [reply io/module_wrt_grid_comp.F90]: Yeah I had to do this to get it to build with fms@2023.04 (and I wasn't able to get some other versions that I tried to work). @junwang-noaa is there an fms version that's best suited to the current fv3atm develop branch?

File: ci/spack.yaml (hunk 2 above)
Comment (AlexanderRichert-NOAA) [reply ci/spack.yaml:12]: Will do.

Spack wants "spack.yaml"

File: ci/CMakeLists.txt (hunk 1 above)
Comment (AlexanderRichert-NOAA) [reply ci/CMakeLists.txt:8]: Done

File: ci/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# This file is used by fv3atm's root CMakeLists.txt when BUILD_TESTING=ON. It is
+# used for CI runs and is based on the ufs-weather-model root CMakeLists.txt. It
+# cannot be built directly, and should not be used for compiling for general R&D
+# or operations.
+
```
Comment (edwardhartnett) [inline ci/CMakeLists.txt:5]: Put your name and date too...

**Comment by edwardhartnett** [comment]
@junwang-noaa can you review this PR? This is needed for both documentation and unit testing work.

File: .github/workflows/GCC.yml [hunk 5]
```diff
@@ -0,0 +1,91 @@
+# This is a CI workflow for the fv3atm project.
+#
+# This workflow builds and tests the fv3atm library using GCC, and it tests
+# different CMake build options.
+#
+# Alex Richert, 6 Dec 2023
+
+name: GCC
```
Comment (junwang-noaa) [inline .github/workflows/GCC.yml:8]: What does GCC stand for?

File: CMakeLists.txt [hunk 6]
```diff
@@ -81,7 +87,7 @@ add_library(fv3atm
   fv3_cap.F90
   module_fv3_config.F90
   module_fcst_grid_comp.F90
-  stochastic_physics/stochastic_physics_wrapper.F90
+  stochastic_physics_wrapper/stochastic_physics_wrapper.F90
```
Comment (junwang-noaa) [inline CMakeLists.txt]: What is this change used for? I don't have stochastic_physics_wrapper/stochastic_physics_wrapper.F90 under fv3atm.

File: io/module_wrt_grid_comp.F90 (hunk 3 above)
Comment (junwang-noaa) [reply io/module_wrt_grid_comp.F90]: @AlexanderRichert-NOAA Maybe you can combine changes in [PR#732](https://github.com/NOAA-EMC/fv3atm/pull/732/files)?

File: .github/workflows/GCC.yml (hunk 5 above)
Comment (edwardhartnett) [reply .github/workflows/GCC.yml:8]: GNU Compiler Collection - it's the GNU compilers.

**Comment by edwardhartnett** [comment]
@AlexanderRichert-NOAA when the UFS is build by spack, is fv3atm one of the packages it builds?

Or does it build fv3atm as part of another package, using the git submodule method to get the fv3atm code?

File: .github/workflows/GCC.yml (hunk 5 above)
Comment (junwang-noaa) [reply .github/workflows/GCC.yml:8]: Sorry, I thought it's something else.

File: CMakeLists.txt (hunk 6 above)
Comment (AlexanderRichert-NOAA) [reply CMakeLists.txt]: I'm not sure why I made that change, so I just reverted it.

**Comment by AlexanderRichert-NOAA** [comment]
The CMakeLists.txt of ufs-weather-model uses `add_subdirectory(FV3)`, where FV3 is the name of the fv3atm repository, which is accessed as a git submodule.

File: io/module_wrt_grid_comp.F90 (hunk 3 above)
Comment (AlexanderRichert-NOAA) [reply io/module_wrt_grid_comp.F90]: done

**Comment by AlexanderRichert-NOAA** [comment]
Latest runs are at https://github.com/AlexanderRichert-NOAA/fv3atm/actions/runs/7414791796

File: io/module_wrt_grid_comp.F90 [hunk 7]
```diff
@@ -29,7 +29,6 @@ module module_wrt_grid_comp
       use mpi
       use esmf
       use fms
-      use mpp_mod, only : mpp_init   ! needed for fms 2023.02
```
Comment (junwang-noaa) [inline io/module_wrt_grid_comp.F90]: The changes are already in develop branch, can you merge?

Comment (AlexanderRichert-NOAA) [reply io/module_wrt_grid_comp.F90]: Done

**Review by junwang-noaa — APPROVED** [review]

**Comment by AlexanderRichert-NOAA** [comment]
@edwardhartnett if you're able to approve this I think it's good to go

**Review by edwardhartnett — APPROVED** [review]

**Comment by AlexanderRichert-NOAA** [comment]
@junwang-noaa can you recommend a second "approved" reviewer (I take it that Ed is not one)? If we can get this merged, I'll follow it up with the addition of the unit test infrastructure, including the first one or two unit tests.

**Comment by junwang-noaa** [comment]
@BrianCurtis-NOAA Dusan is on leave, I added you to review. Please note that other FV3 PRs may need to  be synced when this PR is committed. Thanks

**Review by BrianCurtis-NOAA — APPROVED** [review]
i look forward to seeing how this works!

**Comment by AlexanderRichert-NOAA** [comment]
@junwang-noaa I believe this is ready to be merged
